<h1>Содержание<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><span><a href="#Функции-и-библиотеки" data-toc-modified-id="Функции-и-библиотеки-1"><span class="toc-item-num">1&nbsp;&nbsp;</span>Функции и библиотеки</a></span></li><li><span><a href="#Обзор-данных" data-toc-modified-id="Обзор-данных-2"><span class="toc-item-num">2&nbsp;&nbsp;</span>Обзор данных</a></span></li><li><span><a href="#Предобработка-данных" data-toc-modified-id="Предобработка-данных-3"><span class="toc-item-num">3&nbsp;&nbsp;</span>Предобработка данных</a></span><ul class="toc-item"><li><span><a href="#Наименования-столбцов" data-toc-modified-id="Наименования-столбцов-3.1"><span class="toc-item-num">3.1&nbsp;&nbsp;</span>Наименования столбцов</a></span></li><li><span><a href="#Дублирование-индекса" data-toc-modified-id="Дублирование-индекса-3.2"><span class="toc-item-num">3.2&nbsp;&nbsp;</span>Дублирование индекса</a></span></li><li><span><a href="#Очистка-данных" data-toc-modified-id="Очистка-данных-3.3"><span class="toc-item-num">3.3&nbsp;&nbsp;</span>Очистка данных</a></span></li><li><span><a href="#Обработка-пропусков" data-toc-modified-id="Обработка-пропусков-3.4"><span class="toc-item-num">3.4&nbsp;&nbsp;</span>Обработка пропусков</a></span></li><li><span><a href="#Преобразование-категориальных-признаков-в-численные" data-toc-modified-id="Преобразование-категориальных-признаков-в-численные-3.5"><span class="toc-item-num">3.5&nbsp;&nbsp;</span>Преобразование категориальных признаков в численные</a></span></li><li><span><a href="#Масштабирование-признаков" data-toc-modified-id="Масштабирование-признаков-3.6"><span class="toc-item-num">3.6&nbsp;&nbsp;</span>Масштабирование признаков</a></span></li></ul></li><li><span><a href="#Разделение-исходных-данные-на-обучающую,-валидационную-и-тестовую-выборки" data-toc-modified-id="Разделение-исходных-данные-на-обучающую,-валидационную-и-тестовую-выборки-4"><span class="toc-item-num">4&nbsp;&nbsp;</span>Разделение исходных данные на обучающую, валидационную и тестовую выборки</a></span><ul class="toc-item"><li><span><a href="#Масштабирование-признаков" data-toc-modified-id="Масштабирование-признаков-4.1"><span class="toc-item-num">4.1&nbsp;&nbsp;</span>Масштабирование признаков</a></span></li></ul></li><li><span><a href="#Обучение-модели" data-toc-modified-id="Обучение-модели-5"><span class="toc-item-num">5&nbsp;&nbsp;</span>Обучение модели</a></span><ul class="toc-item"><li><span><a href="#без-учета-дисбаланса" data-toc-modified-id="без-учета-дисбаланса-5.1"><span class="toc-item-num">5.1&nbsp;&nbsp;</span>без учета дисбаланса</a></span></li><li><span><a href="#Устранение-дисбаланса" data-toc-modified-id="Устранение-дисбаланса-5.2"><span class="toc-item-num">5.2&nbsp;&nbsp;</span>Устранение дисбаланса</a></span><ul class="toc-item"><li><span><a href="#Увеличение-выборки" data-toc-modified-id="Увеличение-выборки-5.2.1"><span class="toc-item-num">5.2.1&nbsp;&nbsp;</span>Увеличение выборки</a></span></li><li><span><a href="#Уменьшение-выборки" data-toc-modified-id="Уменьшение-выборки-5.2.2"><span class="toc-item-num">5.2.2&nbsp;&nbsp;</span>Уменьшение выборки</a></span></li></ul></li><li><span><a href="#с-учетом-дисбаланса" data-toc-modified-id="с-учетом-дисбаланса-5.3"><span class="toc-item-num">5.3&nbsp;&nbsp;</span>с учетом дисбаланса</a></span><ul class="toc-item"><li><span><a href="#Для-увеличиной-выборки" data-toc-modified-id="Для-увеличиной-выборки-5.3.1"><span class="toc-item-num">5.3.1&nbsp;&nbsp;</span>Для увеличиной выборки</a></span></li><li><span><a href="#Для-уменьшеной-выборки" data-toc-modified-id="Для-уменьшеной-выборки-5.3.2"><span class="toc-item-num">5.3.2&nbsp;&nbsp;</span>Для уменьшеной выборки</a></span></li></ul></li><li><span><a href="#оптимизация-гиперпараметров-для-моделей-с-лучшими-показателями" data-toc-modified-id="оптимизация-гиперпараметров-для-моделей-с-лучшими-показателями-5.4"><span class="toc-item-num">5.4&nbsp;&nbsp;</span>оптимизация гиперпараметров для моделей с лучшими показателями</a></span><ul class="toc-item"><li><span><a href="#RandomForestClassifier-(увеличенная-выборка)" data-toc-modified-id="RandomForestClassifier-(увеличенная-выборка)-5.4.1"><span class="toc-item-num">5.4.1&nbsp;&nbsp;</span>RandomForestClassifier (увеличенная выборка)</a></span></li><li><span><a href="#GradientBoostingClassifier-(увеличенная-выборка)" data-toc-modified-id="GradientBoostingClassifier-(увеличенная-выборка)-5.4.2"><span class="toc-item-num">5.4.2&nbsp;&nbsp;</span>GradientBoostingClassifier (увеличенная выборка)</a></span></li><li><span><a href="#GradientBoostingClassifier-(уменьшенная-выборка)" data-toc-modified-id="GradientBoostingClassifier-(уменьшенная-выборка)-5.4.3"><span class="toc-item-num">5.4.3&nbsp;&nbsp;</span>GradientBoostingClassifier (уменьшенная выборка)</a></span></li><li><span><a href="#MLPClassifier-(увеличенная-выборка)" data-toc-modified-id="MLPClassifier-(увеличенная-выборка)-5.4.4"><span class="toc-item-num">5.4.4&nbsp;&nbsp;</span>MLPClassifier (увеличенная выборка)</a></span></li></ul></li></ul></li><li><span><a href="#Финальное-тестирование" data-toc-modified-id="Финальное-тестирование-6"><span class="toc-item-num">6&nbsp;&nbsp;</span>Финальное тестирование</a></span></li><li><span><a href="#Заключение" data-toc-modified-id="Заключение-7"><span class="toc-item-num">7&nbsp;&nbsp;</span>Заключение</a></span></li></ul></div>

**Входные данные** —  исторические данные о поведении клиентов и расторжении договоров с банком.

**Дальнейшее использование результатов исследования** — прогноз ухода клиента из банка в ближайшее время с целью планирования мероприятий по его удержанию.

**Цель исследования** 
1. Построение модели с предельно большим значением F1-меры (> 0.59)

2. Измерение AUC-ROC и сравнивнение с F1-мерой 

## Функции и библиотеки

In [68]:
import pandas as pd

import numpy as np

from sklearn.preprocessing import OrdinalEncoder, StandardScaler 

from sklearn.model_selection import train_test_split

from sklearn.tree import DecisionTreeClassifier

from sklearn.ensemble import RandomForestClassifier, GradientBoostingClassifier

from sklearn.linear_model import LogisticRegression

from sklearn.neural_network import MLPClassifier

from sklearn.metrics import accuracy_score, f1_score, roc_auc_score

from sklearn.neighbors import KNeighborsClassifier

from sklearn.utils import shuffle

from sklearn.model_selection import RepeatedStratifiedKFold, RandomizedSearchCV

from scipy.stats import uniform as sp_randFloat

from scipy.stats import randint as sp_randInt

import warnings
warnings.filterwarnings('ignore')

In [2]:
def upsample(features, target):
    features_zeros = features[target == 0]
    features_ones = features[target == 1]
    target_zeros = target[target == 0]
    target_ones = target[target == 1]
    
    repeat = int(round(len(target_zeros)/len(target_ones),0))

    features_upsampled = pd.concat([features_zeros] + [features_ones] * repeat)
    target_upsampled = pd.concat([target_zeros] + [target_ones] * repeat)
    
    features_upsampled, target_upsampled = shuffle(
        features_upsampled, target_upsampled, random_state=12345)
    
    return  features_upsampled, target_upsampled

In [3]:
def downsample(features, target):
      
    features_zeros = features[target == 0]
    features_ones = features[target == 1]
    target_zeros = target[target == 0]
    target_ones = target[target == 1]
    
    fraction = round(len(target_ones)/len(target_zeros),2)

    features_downsampled = pd.concat(
        [features_zeros.sample(frac=fraction, random_state=12345)] + [features_ones])
    target_downsampled = pd.concat(
        [target_zeros.sample(frac=fraction, random_state=12345)] + [target_ones])
    
    features_downsampled, target_downsampled = shuffle(
        features_downsampled, target_downsampled, random_state=12345)
    
    return features_downsampled, target_downsampled

## Обзор данных

**Признаки**
- RowNumber — индекс строки в данных
- CustomerId — уникальный идентификатор клиента
- Surname — фамилия
- CreditScore — кредитный рейтинг
- Geography — страна проживания
- Gender — пол
- Age — возраст
- Tenure — сколько лет человек является клиентом банка
- Balance — баланс на счёте
- NumOfProducts — количество продуктов банка, используемых клиентом
- HasCrCard — наличие кредитной карты
- IsActiveMember — активность клиента
- EstimatedSalary — предполагаемая зарплата

**Целевой признак**
- Exited — факт ухода клиента

In [4]:
try:
    df = pd.read_csv ('D:/Data for projects/Churn.csv')

except: #метод try/except применяется для открытия блокнота по адресу Практикума
    df = pd.read_csv ('/datasets/Churn.csv')

In [5]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10000 entries, 0 to 9999
Data columns (total 14 columns):
RowNumber          10000 non-null int64
CustomerId         10000 non-null int64
Surname            10000 non-null object
CreditScore        10000 non-null int64
Geography          10000 non-null object
Gender             10000 non-null object
Age                10000 non-null int64
Tenure             9091 non-null float64
Balance            10000 non-null float64
NumOfProducts      10000 non-null int64
HasCrCard          10000 non-null int64
IsActiveMember     10000 non-null int64
EstimatedSalary    10000 non-null float64
Exited             10000 non-null int64
dtypes: float64(3), int64(8), object(3)
memory usage: 1.1+ MB


In [6]:
df.head()

RowNumber  CustomerId   Surname  CreditScore Geography  Gender  Age  \
0          1    15634602  Hargrave          619    France  Female   42   
1          2    15647311      Hill          608     Spain  Female   41   
2          3    15619304      Onio          502    France  Female   42   
3          4    15701354      Boni          699    France  Female   39   
4          5    15737888  Mitchell          850     Spain  Female   43   

   Tenure    Balance  NumOfProducts  HasCrCard  IsActiveMember  \
0     2.0       0.00              1          1               1   
1     1.0   83807.86              1          0               1   
2     8.0  159660.80              3          1               0   
3     1.0       0.00              2          0               0   
4     2.0  125510.82              1          1               1   

   EstimatedSalary  Exited  
0        101348.88       1  
1        112542.58       0  
2        113931.57       1  
3         93826.63       0  
4         79084.10       0

In [7]:
df.nunique()

RowNumber          10000
CustomerId         10000
Surname             2932
CreditScore          460
Geography              3
Gender                 2
Age                   70
Tenure                11
Balance             6382
NumOfProducts          4
HasCrCard              2
IsActiveMember         2
EstimatedSalary     9999
Exited                 2
dtype: int64

In [8]:
df['Exited'].value_counts() #оценим достаточность разнообразия целевого признака для обучения

0    7963
1    2037
Name: Exited, dtype: int64

In [9]:
df['EstimatedSalary'].min()

11.58

In [10]:
df['Surname'].value_counts()

Smith     32
Scott     29
Martin    29
Walker    28
Brown     26
          ..
Haddon     1
Real       1
Gomes      1
Cran       1
Persse     1
Name: Surname, Length: 2932, dtype: int64

In [11]:
df.isna().sum()

RowNumber            0
CustomerId           0
Surname              0
CreditScore          0
Geography            0
Gender               0
Age                  0
Tenure             909
Balance              0
NumOfProducts        0
HasCrCard            0
IsActiveMember       0
EstimatedSalary      0
Exited               0
dtype: int64

In [12]:
len(df[df['Tenure'].isna()]['Tenure'])*100/len(df['Tenure']) #расчет % данных с пропусками в столбце Tenure

9.09

<div style="border:solid blue 2px; padding: 20px">

По результатам обзора данных выявлено:
    
    - не соответствие наименований признаков общепринятым
    - дублирование индекса и индекса строки
    - наличие подозрительно низких значений предполагаемой зарплаты (может быть связано с тем, что банк дла клиента не является основным, что может быть важным признаком)
    - наличие данных не влияющих на целевой признак (фамилия клиента), при условии что данный признак не позволяет установить родственные связи между клиентами
    - пропущенные значения в столбце Tenure

## Предобработка данных

### Наименования столбцов

In [13]:
df = df.rename(columns={'RowNumber': 'row_number', 
                        'CustomerId': 'customer_id', 
                        'Surname':'surname',
                        'CreditScore': 'credit_score',
                        'Geography':'geography',
                        'Gender':'gender',
                        'Age':'age',
                        'Tenure':'tenure',
                        'Balance':'balance',
                        'NumOfProducts':'num_of_products',
                        'HasCrCard':'has_cr_card',
                        'IsActiveMember':'is_active_member',
                        'EstimatedSalary':'estimated_salary',
                        'Exited':'exited'
                       })
df.columns #проверка

Index(['row_number', 'customer_id', 'surname', 'credit_score', 'geography',
       'gender', 'age', 'tenure', 'balance', 'num_of_products', 'has_cr_card',
       'is_active_member', 'estimated_salary', 'exited'],
      dtype='object')

### Дублирование индекса

In [14]:
df = df.drop('row_number',axis=1) #уберем столбец row_number, т.к. он дублирует индекс df

### Очистка данных

In [15]:
df = df.drop(['surname','customer_id'],axis=1) #уберем столбцы surname и customer_id, т.к. содержащиеся в них данные не влияют на целевой признак

In [16]:
df.head()

credit_score geography  gender  age  tenure    balance  num_of_products  \
0           619    France  Female   42     2.0       0.00                1   
1           608     Spain  Female   41     1.0   83807.86                1   
2           502    France  Female   42     8.0  159660.80                3   
3           699    France  Female   39     1.0       0.00                2   
4           850     Spain  Female   43     2.0  125510.82                1   

   has_cr_card  is_active_member  estimated_salary  exited  
0            1                 1         101348.88       1  
1            0                 1         112542.58       0  
2            1                 0         113931.57       1  
3            0                 0          93826.63       0  
4            1                 1          79084.10       0

### Обработка пропусков

При обзоре выявлены пропуски в ~ 9% признаков tenure, поскольку восстановить пропущенные данные не представляется возможным, а после их исключения остается достаточно данных для обучения, исключим их

df = df.dropna()

In [17]:
df = df.fillna(-1)

In [18]:
df['tenure'] = df['tenure'].astype(object)

In [19]:
df['tenure'].dtypes

dtype('O')

In [20]:
df['exited'].value_counts() #оценим достаточность разнообразия целевого признака для обучения после обработки пропусков

0    7963
1    2037
Name: exited, dtype: int64

### Преобразование категориальных признаков в численные

Воспользуемся порядковым кодированием (One-Hot Encoding) для того чтобы не сужать выбор возможных моделей (логистическая регрессия)

In [21]:
encoder = OrdinalEncoder() 

df = pd.DataFrame(encoder.fit_transform(df), columns=df.columns) 

### Масштабирование признаков

df.columns

numeric = ['credit_score', 
           'geography', 
           'gender', 
           'age', 
           'tenure', 
           'balance',
           'num_of_products', 
           'has_cr_card', 
           'is_active_member',
           'estimated_salary']

scaler = StandardScaler()
scaler.fit(df[numeric])
df[numeric] = scaler.transform(df[numeric])

df.head()

## Разделение исходных данные на обучающую, валидационную и тестовую выборки

In [22]:
features = df.drop('exited',axis=1) # признаки 
target = df['exited'] # целевой признак

In [23]:
features, features_test, target, target_test = train_test_split(
    features, target, test_size=0.20, random_state=404) #выделение тестовой выборки

In [24]:
features_train, features_valid, target_train, target_valid = train_test_split(
    features, target, test_size=0.25, random_state=404) #выделение обучающей и валидационной выборки

In [25]:
print(features_train.shape)
print(features_valid.shape)
print(features_test.shape)
print(target_train.shape)
print(target_valid.shape)
print(target_test.shape)

(6000, 10)
(2000, 10)
(2000, 10)
(6000,)
(2000,)
(2000,)


### Масштабирование признаков

In [26]:
pd.options.mode.chained_assignment = None

numeric = features_train.columns

scaler = StandardScaler()
scaler.fit(features_train[numeric])

features_train[numeric] = scaler.transform(features_train[numeric])
features_valid[numeric] = scaler.transform(features_valid[numeric])
features_test[numeric] = scaler.transform(features_test[numeric])

In [27]:
features_train.head()

credit_score  geography    gender       age    tenure   balance  \
2241      0.311431   0.289572  0.909418  0.280981 -0.449406  0.480197   
4316      2.066100   1.492780  0.909418  0.471509  0.776246 -0.507497   
4826      1.484671  -0.913636  0.909418 -1.433771  1.695485  1.492325   
8944     -1.038314   1.492780  0.909418  0.947830 -1.062231  1.604158   
2362      0.602146   0.289572  0.909418  0.566773 -0.142993  0.671910   

      num_of_products  has_cr_card  is_active_member  estimated_salary  
2241         0.805841     0.640913         -1.037711         -1.574566  
4316        -0.924050     0.640913         -1.037711          0.080133  
4826        -0.924050     0.640913          0.963660         -0.193101  
8944         0.805841     0.640913          0.963660          0.627643  
2362        -0.924050     0.640913          0.963660         -1.739688

## Обучение модели

### без учета дисбаланса

In [28]:
best_model_tree = None
best_result_tree = 0
depth = 0

for _ in range(1, 22, 2):
    model = DecisionTreeClassifier(random_state=404, max_depth=_)
    model.fit(features_train, target_train) # обучение модели
    predictions = model.predict(features_valid) 
    
    score = f1_score(target_valid, predictions) # расчет качества модели
    
    if score > best_result_tree:
        best_model_tree = model # сохранение наилучшей модели
        best_result_tree = score #  сохранение наилучшего значения метрики accuracy
        depth = _ # сохранение глубины лучшей модели

print(f'F1-мера наилучшей модели : {round(best_result_tree,4)}, max_depth = {depth}')

F1-мера наилучшей модели : 0.5653, max_depth = 9


In [29]:
best_model_forest = None
best_result_forest = 0
n_estimators = 0

for _ in range(100, 1001, 100):
    model = RandomForestClassifier(random_state=404, n_estimators=_) 
    model.fit(features_train,target_train) # обучение модели
    predictions = model.predict(features_valid)
    score = f1_score(predictions, target_valid) # расчет качества модели
    
    if score > best_result_forest:
        best_model_forest = model # сохранение наилучшей модели
        best_result_forest = score #  сохранение наилучшего значения метрики accuracy
        n_estimators = _ # сохранение количества деревьев лучшей модели

print(f'F1-мера наилучшей модели: {round(best_result_forest,4)}, кол-во деревьев: {n_estimators}')

F1-мера наилучшей модели: 0.5964, кол-во деревьев: 300


In [30]:
best_model_forest = None
best_result_forest = 0
n_estimators = 0

for _ in range(1, 50):
    model = RandomForestClassifier(random_state=404, n_estimators=_) 
    model.fit(features_train,target_train) # обучение модели
    predictions = model.predict(features_valid)
    score = f1_score(predictions, target_valid) # расчет качества модели
    
    if score > best_result_forest:
        best_model_forest = model # сохранение наилучшей модели
        best_result_forest = score #  сохранение наилучшего значения метрики accuracy
        n_estimators = _ # сохранение количества деревьев лучшей модели

print(f'F1-мера наилучшей модели: {round(best_result_forest,4)}, кол-во деревьев: {n_estimators}')

F1-мера наилучшей модели: 0.6012, кол-во деревьев: 43


In [31]:
best_model_log_reg = None
best_result_log_reg = 0
max_iter_ = 0

for _ in range(1,150,10):
    model = LogisticRegression(random_state=404, solver='lbfgs', max_iter=_) 
    model.fit(features_train,target_train) # обучение модели
    predictions = model.predict(features_valid)
    score = f1_score(predictions, target_valid) # расчет качества модели
    
    if score > best_result_log_reg:
        best_model_log_reg = model # сохранение наилучшей модели
        best_result_log_reg = score #  сохранение наилучшего значения метрики accuracy
        max_iter_ = _ # сохранение количества иттераций лучшей модели

print(f'F1-мера наилучшей модели: {round(best_result_log_reg,4)}, количество иттераций: {max_iter_}')

F1-мера наилучшей модели: 0.2031, количество иттераций: 11


In [32]:
best_model_neighbors = None
best_result_neighbors = 0
neighbors = 0

for _ in range(1,10):
    model = KNeighborsClassifier(n_neighbors=_) 
    model.fit(features_train,target_train) # обучение модели
    predictions = model.predict(features_valid)
    score = f1_score(predictions, target_valid) # расчет качества модели
    
    if score > best_result_neighbors:
        best_model_neighbors = model # сохранение наилучшей модели
        best_result_neighbors = score #  сохранение наилучшего значения метрики accuracy
        neighbors = _ # сохранение количества иттераций лучшей модели

print(f'F1-мера наилучшей модели: {round(best_result_neighbors,4)}, количество соседей: {neighbors}')

F1-мера наилучшей модели: 0.5031, количество соседей: 1


In [33]:
best_model_GBC = None
best_result_GBC = 0
depth = 0

for _ in range(1,10):
    model = GradientBoostingClassifier(n_estimators=100, learning_rate=0.1, max_depth=_, random_state=404) 
    model.fit(features_train,target_train) # обучение модели
    predictions = model.predict(features_valid)
    score = f1_score(predictions, target_valid) # расчет качества модели
    
    if score > best_result_GBC:
        best_model_GBC = model # сохранение наилучшей модели
        best_result_GBC = score #  сохранение наилучшего значения метрики accuracy
        depth = _ # сохранение количества иттераций лучшей модели

print(f'F1-мера наилучшей модели: {round(best_result_GBC,4)}, max_depth: {depth}')

F1-мера наилучшей модели: 0.6012, max_depth: 7


In [34]:
best_model_MLP = None
best_result_MLP = 0
max_iter_ = 0

for _ in range(1,400,100):
    model =MLPClassifier(random_state=1, max_iter=_)
    model.fit(features_train,target_train) # обучение модели
    predictions = model.predict(features_valid)
    score = f1_score(predictions, target_valid) # расчет качества модели
    
    if score > best_result_MLP:
        best_model_MLP = model # сохранение наилучшей модели
        best_result_MLP = score #  сохранение наилучшего значения метрики accuracy
        max_iter_ = _ # сохранение количества иттераций лучшей модели

print(f'F1-мера наилучшей модели: {round(best_result_MLP,4)}, max_iter: {max_iter_}')

F1-мера наилучшей модели: 0.574, max_iter: 301


### Устранение дисбаланса

#### Увеличение выборки

In [35]:
features_upsampled, target_upsampled = upsample(features_train, target_train)

In [36]:
target_upsampled.value_counts()

0.0    4812
1.0    4752
Name: exited, dtype: int64

#### Уменьшение выборки

In [37]:
features_downsampled, target_downsampled = downsample(features_train, target_train)

In [38]:
target_downsampled.value_counts()

0.0    1203
1.0    1188
Name: exited, dtype: int64

### с учетом дисбаланса

#### Для увеличиной выборки

In [39]:
best_model_tree_ups = None
best_result_tree_ups = 0
depth = 0

for _ in range(1,30):
    model = DecisionTreeClassifier(random_state=404, max_depth=_)
    model.fit(features_upsampled, target_upsampled) # обучение модели
    predictions = model.predict(features_valid) 
    
    score = f1_score(target_valid, predictions) # расчет качества модели
    
    if score > best_result_tree_ups:
        best_model_tree_ups = model # сохранение наилучшей модели
        best_result_tree_ups = score #  сохранение наилучшего значения метрики accuracy
        depth = _ # сохранение глубины лучшей модели

print(f'F1-мера наилучшей модели : {round(best_result_tree_ups,4)}, max_depth = {depth}')

F1-мера наилучшей модели : 0.5797, max_depth = 5


In [40]:
best_model_forest_ups = None
best_result_forest_ups = 0
n_estimators = 0

for _ in range(1, 50):
    model = RandomForestClassifier(random_state=404, n_estimators=_) 
    model.fit(features_upsampled, target_upsampled) # обучение модели
    predictions = model.predict(features_valid)
    score = f1_score(predictions, target_valid) # расчет качества модели
    
    if score > best_result_forest_ups:
        best_model_forest_ups = model # сохранение наилучшей модели
        best_result_forest_ups = score #  сохранение наилучшего значения метрики accuracy
        n_estimators = _ # сохранение количества деревьев лучшей модели

print(f'F1-мера наилучшей модели: {round(best_result_forest_ups,4)}, кол-во деревьев: {n_estimators}')

F1-мера наилучшей модели: 0.6303, кол-во деревьев: 34


In [41]:
best_model_log_reg_ups = None
best_result_log_reg_ups = 0
max_iter_ = 0

for _ in range(1,150,10):
    model = LogisticRegression(random_state=404, solver='lbfgs', max_iter=_) 
    model.fit(features_upsampled, target_upsampled) # обучение модели
    predictions = model.predict(features_valid)
    score = f1_score(predictions, target_valid) # расчет качества модели
    
    if score > best_result_log_reg_ups:
        best_model_log_reg_ups = model # сохранение наилучшей модели
        best_result_log_reg_ups = score #  сохранение наилучшего значения метрики accuracy
        max_iter_ = _ # сохранение количества иттераций лучшей модели

print(f'F1-мера наилучшей модели: {round(best_result_log_reg_ups,4)}, количество иттераций: {max_iter_}')

F1-мера наилучшей модели: 0.5026, количество иттераций: 11


In [42]:
best_model_neighbors_ups = None
best_result_neighbors_ups = 0
neighbors = 0

for _ in range(1,10):
    model = KNeighborsClassifier(n_neighbors=_) 
    model.fit(features_upsampled, target_upsampled) # обучение модели
    predictions = model.predict(features_valid)
    score = f1_score(predictions, target_valid) # расчет качества модели
    
    if score > best_result_neighbors_ups:
        best_model_neighbors_ups = model # сохранение наилучшей модели
        best_result_neighbors_ups = score #  сохранение наилучшего значения метрики accuracy
        neighbors = _ # сохранение количества иттераций лучшей модели

print(f'F1-мера наилучшей модели: {round(best_result_neighbors_ups,4)}, количество соседей: {neighbors}')

F1-мера наилучшей модели: 0.5502, количество соседей: 8


In [43]:
best_model_GBC_ups = None
best_result_GBC_ups = 0
depth = 0

for _ in range(1,10):
    model = GradientBoostingClassifier(n_estimators=100, learning_rate=0.1, max_depth=_, random_state=404) 
    model.fit(features_upsampled, target_upsampled) # обучение модели
    predictions = model.predict(features_valid)
    score = f1_score(predictions, target_valid) # расчет качества модели
    
    if score > best_result_GBC_ups:
        best_model_GBC_ups = model # сохранение наилучшей модели
        best_result_GBC_ups = score #  сохранение наилучшего значения метрики accuracy
        depth = _ # сохранение количества иттераций лучшей модели

print(f'F1-мера наилучшей модели: {round(best_result_GBC_ups,4)}, max_depth: {depth}')

F1-мера наилучшей модели: 0.6437, max_depth: 5


In [44]:
best_model_MLP_ups = None
best_result_MLP_ups = 0
max_iter_ = 0

for _ in range(1,400,100):
    model =MLPClassifier(random_state=404, max_iter=_)
    model.fit(features_upsampled, target_upsampled) # обучение модели
    predictions = model.predict(features_valid)
    score = f1_score(predictions, target_valid) # расчет качества модели
    
    if score > best_result_MLP_ups:
        best_model_MLP_ups = model # сохранение наилучшей модели
        best_result_MLP_ups = score #  сохранение наилучшего значения метрики accuracy
        max_iter_ = _ # сохранение количества иттераций лучшей модели

print(f'F1-мера наилучшей модели: {round(best_result_MLP_ups,4)}, max_iter: {max_iter_}')

F1-мера наилучшей модели: 0.6154, max_iter: 101


#### Для уменьшеной выборки

In [45]:
best_model_tree_dws = None
best_result_tree_dws = 0
depth = 0

for _ in range(1,30):
    model = DecisionTreeClassifier(random_state=404, max_depth=_)
    model.fit(features_downsampled, target_downsampled) # обучение модели
    predictions = model.predict(features_valid) 
    
    score = f1_score(target_valid, predictions) # расчет качества модели
    
    if score > best_result_tree_dws:
        best_model_tree_dws = model # сохранение наилучшей модели
        best_result_tree_dws = score #  сохранение наилучшего значения метрики accuracy
        depth = _ # сохранение глубины лучшей модели

print(f'F1-мера наилучшей модели : {round(best_result_tree_dws,4)}, max_depth = {depth}')

F1-мера наилучшей модели : 0.592, max_depth = 6


In [46]:
best_model_forest_dws = None
best_result_forest_dws = 0
n_estimators = 0

for _ in range(1, 50):
    model = RandomForestClassifier(random_state=404, n_estimators=_) 
    model.fit(features_downsampled, target_downsampled) # обучение модели
    predictions = model.predict(features_valid)
    score = f1_score(predictions, target_valid) # расчет качества модели
    
    if score > best_result_forest_dws:
        best_model_forest_dws = model # сохранение наилучшей модели
        best_result_forest_dws = score #  сохранение наилучшего значения метрики accuracy
        n_estimators = _ # сохранение количества деревьев лучшей модели

print(f'F1-мера наилучшей модели: {round(best_result_forest_dws,4)}, кол-во деревьев: {n_estimators}')

F1-мера наилучшей модели: 0.5955, кол-во деревьев: 48


In [47]:
best_model_log_reg_dws = None
best_result_log_reg_dws = 0
max_iter_ = 0

for _ in range(1,150,10):
    model = LogisticRegression(random_state=404, solver='lbfgs', max_iter=_) 
    model.fit(features_downsampled, target_downsampled) # обучение модели
    predictions = model.predict(features_valid)
    score = f1_score(predictions, target_valid) # расчет качества модели
    
    if score > best_result_log_reg_dws:
        best_model_log_reg_dws = model # сохранение наилучшей модели
        best_result_log_reg_dws = score #  сохранение наилучшего значения метрики accuracy
        max_iter_ = _ # сохранение количества иттераций лучшей модели

print(f'F1-мера наилучшей модели: {round(best_result_log_reg_dws,4)}, количество иттераций: {max_iter_}')

F1-мера наилучшей модели: 0.5065, количество иттераций: 11


In [48]:
best_model_neighbors_dws = None
best_result_neighbors_dws = 0
neighbors = 0

for _ in range(1,10):
    model = KNeighborsClassifier(n_neighbors=_) 
    model.fit(features_downsampled, target_downsampled) # обучение модели
    predictions = model.predict(features_valid)
    score = f1_score(predictions, target_valid) # расчет качества модели
    
    if score > best_result_neighbors_dws:
        best_model_neighbors_dws = model # сохранение наилучшей модели
        best_result_neighbors_dws = score #  сохранение наилучшего значения метрики accuracy
        neighbors = _ # сохранение количества иттераций лучшей модели

print(f'F1-мера наилучшей модели: {round(best_result_neighbors_dws,4)}, количество соседей: {neighbors}')

F1-мера наилучшей модели: 0.555, количество соседей: 8


In [49]:
best_model_GBC_dws = None
best_result_GBC_dws = 0
depth = 0

for _ in range(1,10):
    model = GradientBoostingClassifier(n_estimators=100, learning_rate=0.1, max_depth=_, random_state=404) 
    model.fit(features_downsampled, target_downsampled) # обучение модели
    predictions = model.predict(features_valid)
    score = f1_score(predictions, target_valid) # расчет качества модели
    
    if score > best_result_GBC_dws:
        best_model_GBC_dws = model # сохранение наилучшей модели
        best_result_GBC_dws = score #  сохранение наилучшего значения метрики accuracy
        depth = _ # сохранение количества иттераций лучшей модели

print(f'F1-мера наилучшей модели: {round(best_result_GBC_dws,4)}, max_depth: {depth}')

F1-мера наилучшей модели: 0.6121, max_depth: 2


In [50]:
best_model_MLP_dws = None
best_result_MLP_dws = 0
max_iter_ = 0

for _ in range(1,400,100):
    model =MLPClassifier(random_state=404, max_iter=_)
    model.fit(features_downsampled, target_downsampled) # обучение модели
    predictions = model.predict(features_valid)
    score = f1_score(predictions, target_valid) # расчет качества модели
    
    if score > best_result_MLP_dws:
        best_model_MLP_dws = model # сохранение наилучшей модели
        best_result_MLP_dws = score #  сохранение наилучшего значения метрики accuracy
        max_iter_ = _ # сохранение количества иттераций лучшей модели

print(f'F1-мера наилучшей модели: {round(best_result_MLP_dws,4)}, max_iter: {max_iter_}')

F1-мера наилучшей модели: 0.6007, max_iter: 201


### оптимизация гиперпараметров для моделей с лучшими показателями

Выберем модели с лучшими показателями метрики f1

In [51]:
print(round(best_result_tree,4),'\n',
      round(best_result_forest,4),'\n',
      round(best_result_log_reg,4),'\n',
      round(best_result_neighbors,4),'\n',
      round(best_result_GBC,4),'\n',
      round(best_result_MLP,4))

0.5653 
 0.6012 
 0.2031 
 0.5031 
 0.6012 
 0.574


In [52]:
print(round(best_result_tree_ups,4),'\n',
      round(best_result_forest_ups,4),'\n',
      round(best_result_log_reg_ups,4),'\n',
      round(best_result_neighbors_ups,4),'\n',
      round(best_result_GBC_ups,4),'\n',
      round(best_result_MLP_ups,4))

0.5797 
 0.6303 
 0.5026 
 0.5502 
 0.6437 
 0.6154


In [53]:
print(round(best_result_tree_dws,4),'\n',
      round(best_result_forest_dws,4),'\n',
      round(best_result_log_reg_dws,4),'\n',
      round(best_result_neighbors_dws,4),'\n',
      round(best_result_GBC_dws,4),'\n',
      round(best_result_MLP_dws,4))

0.592 
 0.5955 
 0.5065 
 0.555 
 0.6121 
 0.6007


Согласно условиям кейса минимально допустимое значение метрики F1 соответствует 0,59

Данному условию удовлетворяют следующие модели:
- RandomForestClassifier, обученная на не сбалансированной выборке (best_result_forest)
- GradientBoostingClassifier, обученная на не сбалансированной выборке (best_result_GBC)

- RandomForestClassifier, обученная на увеличенной выборке (best_result_forest_ups)
- GradientBoostingClassifier, обученная на увеличенной выборке (best_result_GBC_ups)
- MLPClassifier, обученная на увеличенной выборке (best_result_MLP_ups)

- RandomForestClassifier, обученная на уменьшенной выборке (best_result_forest_dws)
- GradientBoostingClassifier, обученная на уменьшенной выборке (best_result_GBC_dws)
- MLPClassifier, обученная на уменьшенной выборке (best_result_MLP_dws)

проведем для данных моделей оптимизацию гиперпараметров, расчитаем и сравним метрики (f1, AUC ROC), выберем модель с лучшими показателями

#### RandomForestClassifier (увеличенная выборка)

In [69]:
model = RandomForestClassifier()

n_estimators = [int(x) for x in np.linspace(start = 100, stop = 1000, num = 10)]
max_features = ['log2', 'sqrt']
max_depth = [int(x) for x in np.linspace(start = 1, stop = 15, num = 15)]
min_samples_split = [int(x) for x in np.linspace(start = 2, stop = 50, num = 10)]
min_samples_leaf = [int(x) for x in np.linspace(start = 2, stop = 50, num = 10)]
bootstrap = [True, False]
parameters = {'n_estimators': n_estimators,
               'max_features': max_features,
               'max_depth': max_depth,
               'min_samples_split': min_samples_split,
               'min_samples_leaf': min_samples_leaf,
               'bootstrap': bootstrap}

randm_src = RandomizedSearchCV(estimator=model, param_distributions = parameters,
                               cv = 5, n_iter = 10, n_jobs=-1, scoring='f1')
randm_src.fit(features_upsampled, target_upsampled)

print(" Results from Random Search " )
print("\n The best estimator across ALL searched params:\n", randm_src.best_estimator_)
print("\n The best score across ALL searched params:\n", randm_src.best_score_)
print("\n The best parameters across ALL searched params:\n", randm_src.best_params_)

 Results from Random Search 

 The best estimator across ALL searched params:
 RandomForestClassifier(bootstrap=True, class_weight=None, criterion='gini',
                       max_depth=12, max_features='log2', max_leaf_nodes=None,
                       min_impurity_decrease=0.0, min_impurity_split=None,
                       min_samples_leaf=12, min_samples_split=50,
                       min_weight_fraction_leaf=0.0, n_estimators=400,
                       n_jobs=None, oob_score=False, random_state=None,
                       verbose=0, warm_start=False)

 The best score across ALL searched params:
 0.8031263538779001

 The best parameters across ALL searched params:
 {'n_estimators': 400, 'min_samples_split': 50, 'min_samples_leaf': 12, 'max_features': 'log2', 'max_depth': 12, 'bootstrap': True}


In [71]:
model = RandomForestClassifier(bootstrap=True, class_weight=None, criterion='gini',
                       max_depth=12, max_features='log2', max_leaf_nodes=None,
                       min_impurity_decrease=0.0, min_impurity_split=None,
                       min_samples_leaf=12, min_samples_split=50,
                       min_weight_fraction_leaf=0.0, n_estimators=400,
                       n_jobs=None, oob_score=False, random_state=None,
                       verbose=0, warm_start=False) 

model.fit(features_upsampled, target_upsampled) # обучение модели

predictions = model.predict(features_valid)

print('метрика f1:',f1_score(predictions, target_valid)) 

probabilities_valid = model.predict_proba(features_valid)
probabilities_one_valid = probabilities_valid[:, 1]

print ('метрика AUC_ROC:',roc_auc_score(target_valid, probabilities_one_valid))

метрика f1: 0.6287553648068669
метрика AUC_ROC: 0.8699047312251313


In [66]:
model = RandomForestClassifier(random_state=404, n_estimators=34) 

model.fit(features_upsampled, target_upsampled) # обучение модели

predictions = model.predict(features_valid)

print('метрика f1:',f1_score(predictions, target_valid)) 

probabilities_valid = model.predict_proba(features_valid)
probabilities_one_valid = probabilities_valid[:, 1]

print ('метрика AUC_ROC:',roc_auc_score(target_valid, probabilities_one_valid))

метрика f1: 0.6303191489361702
метрика AUC_ROC: 0.8530567138536428


Оптимизация гиперпараметров с использованием RandomizedSearchCV не дала результатов

#### GradientBoostingClassifier (увеличенная выборка)

In [54]:
model = GradientBoostingClassifier()

parameters = {'learning_rate': sp_randFloat(),
              'subsample'    : sp_randFloat(),
              'n_estimators' : sp_randInt(100, 1000),
              'max_depth'    : sp_randInt(1, 10)}

randm_src = RandomizedSearchCV(estimator=model, param_distributions = parameters,
                               cv = 5, n_iter = 10, n_jobs=-1, scoring='f1')
randm_src.fit(features_upsampled, target_upsampled)

print(" Results from Random Search " )
print("\n The best estimator across ALL searched params:\n", randm_src.best_estimator_)
print("\n The best score across ALL searched params:\n", randm_src.best_score_)
print("\n The best parameters across ALL searched params:\n", randm_src.best_params_)

 Results from Random Search 

 The best estimator across ALL searched params:
 GradientBoostingClassifier(criterion='friedman_mse', init=None,
                           learning_rate=0.6230063036968986, loss='deviance',
                           max_depth=9, max_features=None, max_leaf_nodes=None,
                           min_impurity_decrease=0.0, min_impurity_split=None,
                           min_samples_leaf=1, min_samples_split=2,
                           min_weight_fraction_leaf=0.0, n_estimators=199,
                           n_iter_no_change=None, presort='auto',
                           random_state=None, subsample=0.6586929288265204,
                           tol=0.0001, validation_fraction=0.1, verbose=0,
                           warm_start=False)

 The best score across ALL searched params:
 0.9561381118082813

 The best parameters across ALL searched params:
 {'learning_rate': 0.6230063036968986, 'max_depth': 9, 'n_estimators': 199, 'subsample': 0.658692928

In [55]:
model = GradientBoostingClassifier (criterion='friedman_mse', init=None,
                           learning_rate=0.35065398372363465, loss='deviance',
                           max_depth=7, max_features=None, max_leaf_nodes=None,
                           min_impurity_decrease=0.0, min_impurity_split=None,
                           min_samples_leaf=1, min_samples_split=2,
                           min_weight_fraction_leaf=0.0, n_estimators=267,
                           n_iter_no_change=None, presort='auto',
                           random_state=None, subsample=0.865160193889672,
                           tol=0.0001, validation_fraction=0.1, verbose=0,
                           warm_start=False) 

model.fit(features_upsampled, target_upsampled) # обучение модели

predictions = model.predict(features_valid)

print('метрика f1:',f1_score(predictions, target_valid)) 

probabilities_valid = model.predict_proba(features_valid)
probabilities_one_valid = probabilities_valid[:, 1]

print ('метрика AUC_ROC:',roc_auc_score(target_valid, probabilities_one_valid))

метрика f1: 0.5888738127544098
метрика AUC_ROC: 0.8465901295106514


In [56]:
model = GradientBoostingClassifier(n_estimators=100, 
                                   learning_rate=0.1, 
                                   max_depth=3, 
                                   random_state=404) 

model.fit(features_upsampled, target_upsampled) # обучение модели

predictions = model.predict(features_valid)

print('метрика f1:',f1_score(predictions, target_valid)) 

probabilities_valid = model.predict_proba(features_valid)
probabilities_one_valid = probabilities_valid[:, 1]

print ('метрика AUC_ROC:',roc_auc_score(target_valid, probabilities_one_valid))

метрика f1: 0.6360975609756098
метрика AUC_ROC: 0.8783622957567516


<div style="border:solid blue 2px; padding: 20px">

Оптимизация гиперпараметров с использованием RandomizedSearchCV не дала результатов

#### GradientBoostingClassifier (уменьшенная выборка)

In [57]:
model = GradientBoostingClassifier()

parameters = {'learning_rate': sp_randFloat(),
              'subsample'    : sp_randFloat(),
              'n_estimators' : sp_randInt(100, 1000),
              'max_depth'    : sp_randInt(1, 10)}

randm_src = RandomizedSearchCV(estimator=model, param_distributions = parameters,
                               cv = 5, n_iter = 10, n_jobs=-1, scoring='f1')
randm_src.fit(features_downsampled, target_downsampled)

print(" Results from Random Search " )
print("\n The best estimator across ALL searched params:\n", randm_src.best_estimator_)
print("\n The best score across ALL searched params:\n", randm_src.best_score_)
print("\n The best parameters across ALL searched params:\n", randm_src.best_params_)

 Results from Random Search 

 The best estimator across ALL searched params:
 GradientBoostingClassifier(criterion='friedman_mse', init=None,
                           learning_rate=0.7653978532398514, loss='deviance',
                           max_depth=1, max_features=None, max_leaf_nodes=None,
                           min_impurity_decrease=0.0, min_impurity_split=None,
                           min_samples_leaf=1, min_samples_split=2,
                           min_weight_fraction_leaf=0.0, n_estimators=968,
                           n_iter_no_change=None, presort='auto',
                           random_state=None, subsample=0.7713839878881709,
                           tol=0.0001, validation_fraction=0.1, verbose=0,
                           warm_start=False)

 The best score across ALL searched params:
 0.7406819417807075

 The best parameters across ALL searched params:
 {'learning_rate': 0.7653978532398514, 'max_depth': 1, 'n_estimators': 968, 'subsample': 0.771383987

In [58]:
model = GradientBoostingClassifier(criterion='friedman_mse', init=None,
                           learning_rate=0.03958702071337883, loss='deviance',
                           max_depth=1, max_features=None, max_leaf_nodes=None,
                           min_impurity_decrease=0.0, min_impurity_split=None,
                           min_samples_leaf=1, min_samples_split=2,
                           min_weight_fraction_leaf=0.0, n_estimators=728,
                           n_iter_no_change=None, presort='auto',
                           random_state=None, subsample=0.12149086794442532,
                           tol=0.0001, validation_fraction=0.1, verbose=0,
                           warm_start=False) 

model.fit(features_downsampled, target_downsampled) # обучение модели

predictions = model.predict(features_valid)

print('метрика f1:',f1_score(predictions, target_valid)) 

probabilities_valid = model.predict_proba(features_valid)
probabilities_one_valid = probabilities_valid[:, 1]

print ('метрика AUC_ROC:',roc_auc_score(target_valid, probabilities_one_valid))

метрика f1: 0.6035283194057567
метрика AUC_ROC: 0.8549850564633034


In [59]:
model = GradientBoostingClassifier(n_estimators=100, 
                                   learning_rate=0.1, 
                                   max_depth=3, 
                                   random_state=404) 

model.fit(features_downsampled, target_downsampled) # обучение модели

predictions = model.predict(features_valid)

print('метрика f1:',f1_score(predictions, target_valid)) 

probabilities_valid = model.predict_proba(features_valid)
probabilities_one_valid = probabilities_valid[:, 1]

print ('метрика AUC_ROC:',roc_auc_score(target_valid, probabilities_one_valid))

метрика f1: 0.6099815157116452
метрика AUC_ROC: 0.8724549755116894


<div style="border:solid blue 2px; padding: 20px">

Оптимизация гиперпараметров с использованием RandomizedSearchCV не дала результатов

#### MLPClassifier (увеличенная выборка)

In [60]:
import scipy.stats
import itertools
from sklearn.metrics import fbeta_score, make_scorer
from sklearn.metrics import precision_score
from sklearn.model_selection import StratifiedKFold

model = MLPClassifier()

hidden_layer_sizes = ([x for x in itertools.product((10, 20, 30, 40, 50, 53, 73, 100), repeat=1)] + \
                       [x for x in itertools.product((10, 20, 30, 40, 50, 100), repeat=2)] )
                       #[((len(clean_df.columns))+1,)] )]
#print(hidden_layer_sizes)
alpha = [0.01, 0.1, 1, 10, 100]
activation = ['identity', 'logistic', 'tanh', 'relu']
solver = ['lbfgs', 'sgd', 'adam']
learning_rate = ['constant', 'invscaling', 'adaptive']

parameters = {'hidden_layer_sizes': hidden_layer_sizes,
               'alpha': alpha,
              'activation' : activation,
              'solver': solver,
              'learning_rate': learning_rate}


randm_src = RandomizedSearchCV(estimator=model, param_distributions = parameters,
                               n_iter=100, cv=5, n_jobs=-1, verbose=3, scoring='f1')
randm_src.fit(features_downsampled, target_downsampled)

print(" Results from Random Search " )
print("\n The best estimator across ALL searched params:\n", randm_src.best_estimator_)
print("\n The best score across ALL searched params:\n", randm_src.best_score_)
print("\n The best parameters across ALL searched params:\n", randm_src.best_params_)

Fitting 5 folds for each of 100 candidates, totalling 500 fits


[Parallel(n_jobs=-1)]: Using backend LokyBackend with 4 concurrent workers.
[Parallel(n_jobs=-1)]: Done  24 tasks      | elapsed:    9.6s
[Parallel(n_jobs=-1)]: Done 120 tasks      | elapsed:  1.2min
[Parallel(n_jobs=-1)]: Done 280 tasks      | elapsed:  2.7min
[Parallel(n_jobs=-1)]: Done 500 out of 500 | elapsed:  5.2min finished


 Results from Random Search 

 The best estimator across ALL searched params:
 MLPClassifier(activation='relu', alpha=10, batch_size='auto', beta_1=0.9,
              beta_2=0.999, early_stopping=False, epsilon=1e-08,
              hidden_layer_sizes=(100,), learning_rate='invscaling',
              learning_rate_init=0.001, max_iter=200, momentum=0.9,
              n_iter_no_change=10, nesterovs_momentum=True, power_t=0.5,
              random_state=None, shuffle=True, solver='lbfgs', tol=0.0001,
              validation_fraction=0.1, verbose=False, warm_start=False)

 The best score across ALL searched params:
 0.7592437260991816

 The best parameters across ALL searched params:
 {'solver': 'lbfgs', 'learning_rate': 'invscaling', 'hidden_layer_sizes': (100,), 'alpha': 10, 'activation': 'relu'}


In [61]:
model =MLPClassifier(activation='tanh', alpha=10, batch_size='auto', beta_1=0.9,
              beta_2=0.999, early_stopping=False, epsilon=1e-08,
              hidden_layer_sizes=(10, 30), learning_rate='constant',
              learning_rate_init=0.001, max_iter=200, momentum=0.9,
              n_iter_no_change=10, nesterovs_momentum=True, power_t=0.5,
              random_state=None, shuffle=True, solver='lbfgs', tol=0.0001,
              validation_fraction=0.1, verbose=False, warm_start=False)

model.fit(features_upsampled, target_upsampled) # обучение модели

predictions = model.predict(features_valid)

print('метрика f1:',f1_score(predictions, target_valid)) 

probabilities_valid = model.predict_proba(features_valid)
probabilities_one_valid = probabilities_valid[:, 1]

print ('метрика AUC_ROC:',roc_auc_score(target_valid, probabilities_one_valid))

метрика f1: 0.6190028222013171
метрика AUC_ROC: 0.8635574565563648


In [62]:
model =MLPClassifier(random_state=404, max_iter=101)

model.fit(features_upsampled, target_upsampled) # обучение модели

predictions = model.predict(features_valid)

print('метрика f1:',f1_score(predictions, target_valid)) 

probabilities_valid = model.predict_proba(features_valid)
probabilities_one_valid = probabilities_valid[:, 1]

print ('метрика AUC_ROC:',roc_auc_score(target_valid, probabilities_one_valid))

метрика f1: 0.6153846153846153
метрика AUC_ROC: 0.8604121552696101


<div style="border:solid blue 2px; padding: 20px">

Оптимизация гиперпараметров с использованием RandomizedSearchCV показала возможность увеличения качества модели

## Финальное тестирование

In [72]:
predictions = best_model_forest_ups.predict(features_test)

probabilities_valid =  best_model_forest_ups.predict_proba(features_test)
probabilities_one_valid = probabilities_valid[:, 1]

print('метрика f1:',f1_score(predictions, target_test)) 
print ('метрика AUC_ROC:',roc_auc_score(target_test, probabilities_one_valid))

метрика f1: 0.5855614973262031
метрика AUC_ROC: 0.8454422392818757


In [63]:
predictions = best_model_GBC_ups.predict(features_test)

probabilities_valid = best_model_GBC_ups.predict_proba(features_test)
probabilities_one_valid = probabilities_valid[:, 1]

print('метрика f1:',f1_score(predictions, target_test)) 
print ('метрика AUC_ROC:',roc_auc_score(target_test, probabilities_one_valid))

метрика f1: 0.5955284552845529
метрика AUC_ROC: 0.863593230705577


In [64]:
predictions = best_model_GBC_dws.predict(features_test)

probabilities_valid = best_model_GBC_dws.predict_proba(features_test)
probabilities_one_valid = probabilities_valid[:, 1]

print('метрика f1:',f1_score(predictions, target_test)) 
print ('метрика AUC_ROC:',roc_auc_score(target_test, probabilities_one_valid))

метрика f1: 0.5941747572815534
метрика AUC_ROC: 0.8534338923442932


In [65]:
predictions = best_model_MLP_ups.predict(features_test)

probabilities_valid = best_model_MLP_ups.predict_proba(features_test)
probabilities_one_valid = probabilities_valid[:, 1]

print('метрика f1:',f1_score(predictions, target_test)) 
print ('метрика AUC_ROC:',roc_auc_score(target_test, probabilities_one_valid))

метрика f1: 0.5821596244131456
метрика AUC_ROC: 0.8535133441567688


<div style="border:solid blue 2px; padding: 20px">

**Лучшие значения метрик f1 и AUC_ROC показала модель GradientBoostingClassifier, обученная на увеличенной выборке**

## Заключение

В рамках настоящего исследования:

- проведена предобработка данных, включающая в том числе:
    - преобразование категориальных переменных в численные
    - масштабирование данных
- проведено обучение ряда моделей: 
    - DecisionTreeClassifier
    - RandomForestClassifier
    - LogisticRegression
    - KNeighborsClassifier
    - GradientBoostingClassifier 
    - MLPClassifier
- исследовано качество обученных моделей на метрике F1 для разных гиперпараметров, для:
    - не сбалансированных данных
    - сбалансированных данных (балансировка производилась как способом увеличения так и способом уменьшения выборки)
- показатели выше минимально требуемых достигли следующие модели:
    - RandomForestClassifier, обученная на не сбалансированной выборке (best_result_forest)
    - GradientBoostingClassifier, обученная на не сбалансированной выборке (best_result_GBC)

    - RandomForestClassifier, обученная на увеличенной выборке (best_result_forest_ups)
    - GradientBoostingClassifier, обученная на увеличенной выборке (best_result_GBC_ups)
    - MLPClassifier, обученная на увеличенной выборке (best_result_MLP_ups)

    - RandomForestClassifier, обученная на уменьшенной выборке (best_result_forest_dws)
    - GradientBoostingClassifier, обученная на уменьшенной выборке (best_result_GBC_dws)
    - MLPClassifier, обученная на уменьшенной выборке (best_result_MLP_dws)
- определны модели, показавшие наилучшие показатели по метрике F1, проведена попытка оптимизации гиперпараметров с помощью RandomizedSearchCV:
    - RandomForestClassifier (увеличенная выборка), f1 = 0.630
    - GradientBoostingClassifier (увеличенная выборка), f1 = 0.636
    - GradientBoostingClassifier (уменьшенная выборка), f1 = 0.609
    - MLPClassifier (увеличенная выборка), f1 = 0.619
- расчитаны метрики F1 и AUC ROC для трех моделей, показавших наилучшие результаты
- проведено финальное тестирование моделей